Add dataBinder support to Tooltip and Popover setContent

IDisposable opened the pull request:
Add the ability to set a binder in the **Tooltip** & **Popover** `options` to support data-binding of the created DOM in `setContent`. This allows having binding expressions in the template that will be applied as the tooltip/popover is created during the show event.  This can be used by setting the dataBinder to a function that will apply bindings. Since the DOM elements are not created until the tooltip is shown, the best place for this seems to be set `setContent` method.

An example of the use of this it for knockout binds, here we create a knockout binding that hooks itself up:

``` javascript
     ko.bindingHandlers.popover = {
         init: function (element, valueAccessor, allBindings, viewModel, bindingContext) {
            var options = $.extend({ disabled: false }, ko.toJS(valueAccessor()));
            ko.utils.domNodeDisposal.addDisposeCallback(element, function () {
                $(element).popover('destroy');
            });
         },
         update: function (element, valueAccessor, allBindings, viewModel, bindingContext) {
            var $element = $(element);
            var options = $.extend({ disabled: false }, ko.toJS(valueAccessor()));
            $element.popover('destroy');
            if (options.disabled)
                return;

            options.dataBinder = function ($tip) {
                ko.cleanNode($tip[0]);
                $tip.applyBindings(viewModel);
            };
            $element.popover(options);
         }
     };
```

IDisposable committed:
Added dataBinder support to in setContent

Add the ability to set a binder in the Tooltip options to support data-binding of the created DOM in setContent. This allows having binding expressions in the template that will be applied as the tooltip is created during the show event.

IDisposable committed:
Added dataBinder support to in setContent

Add the ability to set a binder in the Popover options to support data-binding of the created DOM in setContent. This allows having binding expressions in the template that will be applied as the popover is created during the show event.

twbs-savage commented:
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 7ada8fe35c0442a12302f209b1c1f2a47ab7549b
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/48442468

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

cvrebert commented:
What about perhaps firing a "tooltip/popover DOM created" event instead?

cvrebert commented:
(Also, this lacks unit tests)

IDisposable commented:
I really wasn't sure how to do a test for this, not my usual realm.  We initially considered the idea of a DOM created event,but then struggled over when and where in the lifecycle to call it. This seemed the least intrusive place to enable a hook into something like this, but I'm not a client-side dev by default :(

IDisposable committed:
Fix indent.

Lost a blank in the inline editor

twbs-savage commented:
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: e3a1cc9166dca218a8190b4b1613c6d5af951103
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/48444241

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

IDisposable committed:
Add unit test for dataBinder

IDisposable committed:
Fixed typos.

Perhaps writing my very first unit test inside the github editor isn't the smartest idea.

twbs-savage commented:
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: da59e6638e874a7f97e94b43e0f4479b55fe3711
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/48445369

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

IDisposable committed:
Once more, typos

This time I did the edit elsewhere.

twbs-savage commented:
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: ba5973fedf88d759dd37ea08939adfb93770cb9a
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/48446008

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

IDisposable committed:
Feeding the tab-beast some spaces.

Wow, so now I'm fighting indentation rules, gee this is fun.

twbs-savage commented:
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 9521437dabc6e0d77aa92e1b1e6e931d9cd3b045
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/48446497

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

IDisposable committed:
This is not pleasant

How to collapse all these commits eludes me.

IDisposable commented:
How do I collapse these commits into one so my stupidity doesn't hurt anyone else's head?

twbs-savage commented:
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: cbe9437301dd883ed3c428ab6f70cbd0525f9e56
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/48447010

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

IDisposable commented:
I quit. Someone else show me up by making the unit test work... I don't have ANY of this environment available... I can only RUN the code that worked in my original example.

cvrebert commented:
> How do I collapse these commits into one so my stupidity doesn't hurt anyone else's head?

`git rebase -i 03e41b355c51474ebac21ba22f4626f05feba844`
Then change all but the first `pick` to `squash`
More info: http://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#Changing-Multiple-Commit-Messages

cvrebert commented:
You can run the tests locally using `grunt test`. See http://getbootstrap.com/getting-started/#grunt-installing for how to setup Grunt.

IDisposable commented:
Yeah, I'm a pathetic Windows developer and barely can get Git to not screw up my line endings (hence using the Github inplace editor), pretty sure I'm not going to get grunt going :(  I'm too stupid for open source.

kkirsche commented:
You can do it @IDisposable :+1:

First things first: do you have Node.JS installed on your machine? If you do not, Node has a great Windows installer which can be found at [http://nodejs.org/download/](http://nodejs.org/download/).

This will install Node.JS and NPM. In order to get started with Grunt, you'll want to install Grunt's command line interface (CLI) globally using NPM. You may need to run your command shell as Administrator (for Windows) to do this.
```shell
npm install -g grunt-cli
```

This will put the grunt command in your system path, allowing it to be run from any directory. From there we can get you running the tests no problem :) Good luck, you got this!

IDisposable committed:
Simplified tests.

twbs-savage commented:
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 6f21167635aa831437d3ac8348918604b22152c3
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/48564875

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

IDisposable committed:
Fixed name of dataBinder

Seriously, it's the forest, not the trees.

twbs-savage commented:
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 7d38f964c9a0c04d432db208a8ae7ccc63dd4fb3
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/48565757

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

cvrebert commented:
* As stated previously, this should probably use events rather than callbacks for consistency with the rest of Bootstrap.
* While this feature request seems very reasonable to me, given that Bootstrap v3 is currently under a soft feature freeze, I'm not sure whether we should add this. (If we ultimately don't, I will lobby for it to be added in v4 however.)

CC: @hnrch02 @XhmikosR @juthilo for opinions.

hnrch02 commented:
I can only echo what you said, @cvrebert: Definitely should use events instead of callbacks and I'd personally like to see this in v3 since it isn't that big of an addition - but since 3.4 is unlikely, we'd have to introduce it in a patch...

mustangsilverfox commented:
:)

cvrebert commented:
Closing this in favor of #16124, since this PR uses callbacks instead of events and also doesn't merge cleanly anymore.
@IDisposable Thanks for bringing this to our attention!
